File: 剑指Offer39.数组中出现次数超过一半的数字/剑指Offer39.数组中出现次数超过一半的数字.py
'''
Descroption: 剑指 Offer 39. 数组中出现次数超过一半的数字
Author: EmoryHuang
Date: 2021-07-03 15:21:26
解题思路:
方法一：排序后取中位数
由于数字的出现次数超过一半，因此中位数一定是这个数
方法二：哈希表
'''


class Solution:
    # 方法一：排序后取中位数
    def majorityElement(self, nums: List[int]) -> int:
        return sorted(nums)[len(nums)//2]

    # 方法二：哈希表
    # def majorityElement(self, nums: List[int]) -> int:
    #     return max( collections.Counter(nums).items() ,key= lambda x : x[1] )[0]

File: Solution/剑指Offer39.数组中出现次数超过一半的数字/剑指Offer39.数组中出现次数超过一半的数字.java
/*
 * @Description: 剑指 Offer 56 - I. 数组中数字出现的次数
 * @Author: EmoryHuang
 * @Date: 2023-08-02 10:44:20
 * @Method:
 * 哈希表
 */

import java.util.*;

class Solution {
    public int majorityElement(int[] nums) {
        Map<Integer, Integer> hash = new HashMap<>();
        int n = nums.length;
        for (int num : nums) {
            hash.put(num, hash.getOrDefault(num, 0) + 1);
            if (hash.get(num) * 2 > n) {
                return num;
            }
        }
        return -1;
    }
}
